File: compiler/bench/polyeval.c
// expected: -511200
// Evaluate degree-100 polynomial, 100K times (using Horner's method)
// Checksum: sum of (result mod 1000) for each evaluation

#include <stdio.h>

long coeffs[101];

void init_coeffs(void) {
    for (int i = 0; i < 101; i++) {
        coeffs[i] = i + 1;
    }
}

long poly_eval(long x) {
    long result = 0;
    for (int i = 0; i < 101; i++) {
        result = result * x + coeffs[100 - i];
    }
    return result;
}

int main(void) {
    init_coeffs();
    long sum = 0;
    for (int i = 0; i < 100000; i++) {
        long r = poly_eval(i % 1000);
        sum += r % 1000;
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/polyeval.fs
\ expected: -511200
\ Evaluate degree-100 polynomial, 100K times (using Horner's method)
\ Checksum: sum of (result mod 1000) for each evaluation

create coeffs 101 cells allot

: init-coeffs 101 0 do i 1+ coeffs i cells + ! loop ;

: poly-eval ( x -- result )
  0 101 0 do
    over * coeffs 100 i - cells + @ +
  loop swap drop ;

: bench-polyeval ( -- sum )
  init-coeffs
  0
  100000 0 do
    i 1000 mod poly-eval
    1000 mod +
  loop ;

: main bench-polyeval . cr ;
